### ISSUE TITLE
Sales/register and Receiving data removed after print an old receipt

### ISSUE THREAD

**Issue by freemail624** [jekkos/opensourcepos#403]
If you have some data in sales register or receiving and at the same time tried to make a report to preview or reprint an old receipt then went back to complete sales register you will not find your data (the data was cleared)

**Comment by jekkos** [comment]
Version please?

**Comment by freemail624** [comment]
Version 2.4.0

**Comment by jekkos** [comment]
you mean after sale completion? So when you're on the final page?

**Comment by odiea** [comment]
The way I get this to show is. Add an item or 2 to a sale. You can go to takings or detailed sale report. As long as you do not click to view the receipt the cart still has info. After you view the receipt the cart is emptied. So it appears you have to view the relevant receipt.

**Comment by freemail624** [comment]
I mean like @odiea said 

> Add an item or 2 to a sale. You can go to takings or detailed sale report. As long as you do not click to view the receipt the cart still has info. After you view the receipt the cart is emptied

**Comment by daN4cat** [comment]
Yes, the session data is used for cart or receipt. OSPOS was designed like that, so a change means dealing with things in different way like implenting a stack to push and pop.

What about suspend a sale before going to view receipts? That's the stack to push and pop already available.

**Comment by odiea** [comment]
That's what my suggestion was going to be also. The only issue then would be the receivings with no suspended feature.

**Comment by jekkos** [comment]
the cart is indeed cleared out when going to receipt (as it's reloaded with the data about to be shown)

**Comment by daN4cat** [comment]
Yes, receiving is not covered. But any improvement will not be for 2.4.

I think this issue is "works as per design". We inherited this and improving but with 3.0 and beyond.

For cart please use suspend and resume.

@odiea do you mind to create an issue listing what should be improved of receiving?

**Comment by freemail624** [comment]
What about make new receipt_lib to be for viewing and printing the saved receipts?

**Comment by daN4cat** [comment]
Contributions are welcome. If you could add that functionality we are happy to merge it.

**Comment by daN4cat** [comment]
I close this because any improvement won't be for 2.4.

**Comment by daN4cat** [comment]
I reopen this because it could be a future improvement:

- add a suspended for receivings
- add two different means to store data for cart and reprinting of receipts/invoices (current work around is manual push in suspended and pop from suspended
